pyfr/integrators/dual/pseudo/pseudosteppers.py

```diff
@@ -337,3 +337,234 @@ class DualRK45PseudoStepper(DualRKVdH2RPseudoStepper):
     @property
     def _pseudo_stepper_order(self):
         return 4
+
+
+class DualDenseRKPseudoStepper(BaseDualPseudoStepper):
+    @property
+    def _pseudo_stepper_has_lerrest(self):
+        return False
+
+    def step(self, t):
+        add = self._add
+        rhs = self._rhs_with_dts
+        dtau = self._dtau
+
+        r = self._pseudo_stepper_regidx
+
+        # Ensure r0 references the bank containing u(n+1,m)
+        if r[0] != self._idxcurr:
+            r[0], r[1] = r[1], r[0]
+
+        # First stage
+        rhs(t, r[0], r[1])
+
+        # Other stages
+        for i in range(self._pseudo_stepper_nregs - 2):
+            ar = [ar for ars in zip([a*dtau for a in self.a[i]],
+                                    r[1:]) for ar in ars]
+            add(0, r[i + 2], 1, r[0], *ar)
+            rhs(t, r[i + 2], r[i + 2])
+
+        # Final accumulation
+        br = [br for brs in zip([b*dtau for b in self.b[1:]],
+                                r[2:]) for br in brs]
+        add(self.b[0]*dtau, r[1], 1, r[0], *br)
+
+        # Return the index of the bank containing u(n+1,m+1)
+        return r[1], r[0]
+
+
+class DualRKOptPseudoStepper(DualDenseRKPseudoStepper):
+    pseudo_stepper_name = 'rkopt'
+
+    @property
+    def _stepper_nfevals(self):
+        return 7*self.nsteps
+
+    @property
+    def _pseudo_stepper_order(self):
+        return 1
+
+    @property
+    def _pseudo_stepper_nregs(self):
+        return 8
+
+
+class DualRKp1OptPseudoStepper(DualRKOptPseudoStepper):
+    pseudo_stepper_porder = '1'
+
+    a = [
+        [0.2328811281838825],
+        [0.2007437175473038, 0.2007437175473038],
+        [0.2049035121310683, 0.1359607119764239, 0.1577268986576998],
+        [0.2025180067308366, 0.1198054416074228, 0.1389853029773165,
+         0.2052094755795549],
+        [0.2049946817015693, 0.1287767042315417, 0.1276484770941174,
+         0.1884708417499925, 0.2138853585222901],
+        [0.199545958466094, 0.1315602040416435, 0.1132216959163609,
+         0.1671699406016753, 0.1764729760258336, 0.192146045128098]
+    ]
+
+    b = [
+        0.1888425763820983, 0.1267393084745009, 0.1061061571001407,
+        0.1525740388611983, 0.1650271578073516, 0.1178619741653911,
+        0.1428487872093191
+    ]
+
+
+class DualRKp0OptPseudoStepper(DualRKp1OptPseudoStepper):
```

Comment:
Might be cleaner if we put the p thing as a suffix so DualRKOptPseudoStepper_p0.  We'll never use the class directly by name so it should not be an issue.